app/models/sale.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+class Sale < ApplicationRecord
+  belongs_to :seller, class_name: 'User', optional: true
+  belongs_to :client, class_name: 'User'
+  belongs_to :payment_method
+  belongs_to :invoice
+  has_one :subscription, dependent: :destroy
+  has_many :refunds, dependent: :destroy
+  has_many :articles_sales, dependent: :destroy
+  has_many :articles, through: :articles_sales
+  has_many :sales_subscription_offers, dependent: :destroy
+  has_many :subscription_offers, through: :sales_subscription_offers
+
+  accepts_nested_attributes_for :articles_sales
+
+  validates :total_price, presence: true
+
+  def verify
+    self.verified_at = Time.zone.now if verified_at.nil?
+  end
+
+  def generate(duration:, seller:)
+    return false if duration.to_i.negative?
+
+    return false unless generate_sales_subscription_offers duration.to_i
+
+    self.seller = seller
+    create_associated_subscription duration.to_i if duration.to_i.positive?
+    self.total_price = compute_total_price
+    verify if payment_method&.auto_verify
+    generate_invoice
+    true
+  end
+
+  def compute_total_price
+    total = 0
+    articles_sales.each do |rec|
+      total += rec.quantity * Article.find(rec.article_id).price
+    end
+    sales_subscription_offers.each do |rec|
+      total += rec.quantity * SubscriptionOffer.find(rec.subscription_offer.id).price
+    end
+    total
+  end
+
+  def generate_invoice
+    return if invoice
+
+    self.invoice = Invoice.new
+    invoice.generate_from(self)
+  end
+
+  def empty?
+    articles_sales.empty? && sales_subscription_offers.empty?
+  end
+
+  private
+
+  def generate_sales_subscription_offers(duration)
+    subscription_offers = SubscriptionOffer.order(duration: :desc)
+    if subscription_offers.empty?
+      errors.add(:base, 'There are no subscription offers registered!')
+      return false
+    end
+    subscription_offers.each do |offer|
+      break if duration.zero?
+
+      quantity = duration / offer.duration
+      if quantity.positive?
+        sales_subscription_offers.new(subscription_offer_id: offer.id, quantity: quantity)
+        duration -= quantity * offer.duration
+      end
+    end
+    return true if duration.zero?
+
+    errors.add(:base, 'Subscription offers are not exhaustive!')
+    false
+  end
```

Comment:
This function aims to generate the subiscription_offer for a given duration, duration is the number of months the subscription last. Example : an adherent want to pay for 13 months and we have two different offer 12 months is 50€ and 1 month is 5€ then it will create a sales_subscription_offers of 12 months and of 1 months. So we leave the loop when reach 0 because we do not need to continue as we already split the sale into the different offers we have. Lastly, we return false, because it's quite obscure the error process and the throw in my case i'd like just to send a message to the user not crash the website.